modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,133 @@
+package org.tessellation.dag.l0.domain.snapshot.programs
+
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.traverse._
+import cats.syntax.unorderedFoldable._
+
+import org.tessellation.dag.l0.infrastructure.snapshot.{DAGEvent, GlobalSnapshotEvent, StateChannelEvent}
+import org.tessellation.ext.kryo.RefinedSerializer
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.node.shared.domain.event.PeerIdEventCutter
+import org.tessellation.schema.Block
+import org.tessellation.schema.gossip.Ordinal
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.security.Hashed
+import org.tessellation.security.hash._
+
+trait GlobalSnapshotEventCutter[F[_]] extends PeerIdEventCutter[F, GlobalSnapshotEvent]
+
+object GlobalSnapshotEventCutter {
+
+  def make[F[_]: Async: KryoSerializer](maxBinarySizeBytes: Long = 20_000_000L): PeerIdEventCutter[F, GlobalSnapshotEvent] =
+    new PeerIdEventCutter[F, GlobalSnapshotEvent] {
+      type GlobalPeerIdEvents = Map[PeerId, List[(Ordinal, GlobalSnapshotEvent)]]
+
+      def cut(initialEvents: GlobalPeerIdEvents): F[GlobalPeerIdEvents] =
+        initialEvents.tailRecM[F, GlobalPeerIdEvents] { events =>
+          needsCutting(events)
+            .ifM(
+              removeOne(events).map(_.asLeft),
+              events.asRight.pure[F]
+            )
+        }
+
+      def needsCutting(events: GlobalPeerIdEvents): F[Boolean] =
+        events.toBinary.liftTo[F].map(_.length > maxBinarySizeBytes)
+
+      def removeOne(events: GlobalPeerIdEvents): F[GlobalPeerIdEvents] =
+        unpackGlobalPeerIdEvents(events)
+          .flatMap(selectEventToRemove)
+          .map(removeEvent(events, _))
+
+      trait PeerIdEvent {
+        def id: PeerId
+        def fee: Long
+        def hash: Hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent
+      }
+
+      case class PeerIdStateChannelEvent(id: PeerId, event: StateChannelEvent) extends PeerIdEvent {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+        def hash: Hash = event.snapshotBinary.lastSnapshotHash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.asLeft[DAGEvent]
+      }
+
+      case class PeerIdHashedEvent(id: PeerId, event: Hashed[Block]) extends PeerIdEvent {
+        val fee: Long = event.signed.value.transactions.toIterable.map(_.value.fee.value.value).sum
+        def hash: Hash = event.hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.signed.asRight[StateChannelEvent]
+
+        def isMyChild(that: PeerIdHashedEvent): Boolean =
+          that.event.signed.value.parent.contains_(event.ownReference)
+      }
+
+      def unpackGlobalPeerIdEvents(events: GlobalPeerIdEvents): F[(List[PeerIdStateChannelEvent], List[PeerIdHashedEvent])] =
```

Comment:
It might be cleaner to return `Ior[List[PeerIdStateChannelEvent], List[PeerIdHashedEvent]]`